### PR TITLE
Draft of the JavaScript style guide for Ember

### PR DESCRIPTION
**Pull request by twokul**
Related to #9849
- [x] use spaces after the keyword in block statements (#10073)
  - [x] `requireSpaceBeforeKeywords`
  - [x] `requireSpaceAfterKeywords`
- [x] use spaces before block statements (#10095)
  - [x] `requireSpaceBeforeBlockStatements`
- [x] use spaces in ternary conditional statements (#10097)
  - [x] `requireSpacesInConditionalExpression`
- [x] use spaces in functional declarations (#10098)
  - [x] `requireSpacesAfterClosingParenthesisInFunctionDeclaration`
- [x] disallow space before () in call expressions (#10099)
  - [x] `disallowSpacesInCallExpression`
- [x] disallow newline before opening curly brace of all block statements (#10109)
  - [x] `disallowNewlineBeforeBlockStatements`
- [x] disallows placing keywords on a new line (#10110)
  - [x] `disallowKeywordsOnNewLine`
- [ ] disallow block statements from beginning and ending with 2 newlines (#10111)
  - [ ] `disallowPaddingNewlinesInBlocks`
- [x] disallow spaces around round braces when calling a function (#10123)
  - [x] `disallowSpacesInsideRoundBracesInCallExpression`
- [x] disallow empty block statements (#10101)
  - [x] `disallowEmptyBlocks`
- [x] disallow space after opening array square bracket and before closing (#10102)
  - [x] `disallowSpacesInsideArrayBrackets: all`
- [x] disallow space after opening round bracket and before closing (#10103)
- [x] disallow spaces before semicolons (#10084)
  - [x] `disallowSpacesBeforeSemicolons`
- [x] requires blocks to begin and end with a newline (#10100)
  - [x] `requireBlocksOnNewline`
- [x] no leading commas (#10115)
  - [x] `requireCommaBeforeLineBreak`
- [x] no trailing commas
  - [x] `disableTrailingComma`

cc @rwjblue

### CODE REVIEW

**Comment by stavarotti** [comment]
@twokul and @rwjblue Need any help with implementing the items listed above?

File: STYLEGUIDE.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+test('foo-bar', function() {
+});
+```
+
++ no spaces before semicolons.
+
+```javascript
+var foo = {} ;
+```
```
Comment (seanpdoyle) [inline STYLEGUIDE.md]: would it be worth adding `bad` like you do below?

```javascript
// good
var foo = {};
// bad
var foo = {} ;
```

**Comment by twokul** [comment]
@seanpdoyle I think it's pretty clear, no?

File: STYLEGUIDE.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
++ use `push` to add an item to an array.
+
+```javascript
+var foo = [];
+foo.push('bar');
+```
+
+## Variables
+
++ put all declarations on one line.
```
Comment (rwjblue) [inline STYLEGUIDE.md]: `put all non assigning declarations on one line`

File: STYLEGUIDE.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+
+```javascript
+var a, b;
+```
+
++ use multiple lines to declare and assign variables.
```
Comment (rwjblue) [inline STYLEGUIDE.md]: `use a single `var` declaration for each assignment`

File: STYLEGUIDE.md [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+var foo = [1, 2, 3];
+var bar = {a: 'a'};
+```
+
++ no leading commas.
+
+```javascript
+var foo = [
+  1
+  , 2
```
Comment (rwjblue) [inline STYLEGUIDE.md]: This example doesn't follow the rest (all of the rest so far are showing the "right" way to do things but this shows the wrong way).

**Comment by rwjblue** [comment]
Left a few comments, but looks like a great start.  Once these tweaks are made, I think we can merge and continue to iterate.

**Comment by twokul** [comment]
@rwjblue updated per your comments. I think this is ready for another review.

**Comment by rwjblue** [comment]
Looks good.  I'm going to merge and we can continue iterating on master.

**Comment by bcardarella** [comment]
There is no mention of Ember specific styles. Such as preferring:

```javascript
var get = Ember.get
get(this, 'foo');
```

over: `this.get('foo')`

**Comment by rwjblue** [comment]
This styleguide is for Ember itself (not for Ember apps), but I agree we still have more things to add.  A few that come to mind:

* We should definitely mention using `set`/`get` when you want to support CP's or observability in objects accessed by the framework.
* Should also mention not relying on ES5 or Ember specific prototype extensions.

**Comment by sandstrom** [comment]
Neat! Two minuscule thoughts.

1. When possible, things should be expressed in the affirmative. It's easier to read the text that way. E.g. *"Don't use `slice` with `arguments`, use `for` loop."* → *"Use a `for` loop with `arguments` (instead of `slice`)."*

2. The comma-example would probably be padded (although it's about commas and not whitespace, that's how single-line objects are usually written in Ember).

    var foo = [1, 2, 3];
    var bar = { a: 'a' }; // padding added

**Comment by rwjblue** [comment]
@sandstrom - Agreed on both, I look forward to your impending PR :wink:

**Comment by sandstrom** [comment]
@rwjblue Haha, good point, I shouldn't just complain :smile:  https://github.com/emberjs/ember.js/pull/10130

**Pull request merged**